msprime/likelihood.py

```diff
@@ -0,0 +1,195 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Module responsible for computing likelihoods.
+"""
+import math
+import msprime
+import numpy as np
+
+
+def unnormalised_log_mutation_likelihood(arg, theta):
+    # log_likelihood of mutations on a given ARG up to a normalising constant
+    # that depends on the pattern of observed mutations, but not on the ARG
+    # or the mutation rate
+    tables = arg.tables
+    number_of_edges = len(tables.edges)
+    edges_above = {}
+    edge = 0
+    total_material = 0
+    while edge < number_of_edges:
+        child = tables.edges[edge].child
+        right = tables.edges[edge].right
+        left = tables.edges[edge].left
+        parent = tables.edges[edge].parent
+        total_material += ((right - left) *
+                           (tables.nodes[parent].time - tables.nodes[child].time))
+        edges_above[child] = [edge]
+        edge += 1
+        if edge == number_of_edges:
+            break
+        while tables.edges[edge].child == child:
+            right = tables.edges[edge].right
+            left = tables.edges[edge].left
+            parent = tables.edges[edge].parent
+            total_material += ((right - left) *
+                               (tables.nodes[parent].time - tables.nodes[child].time))
+            edges_above[child].append(edge)
+            edge += 1
+            if edge == number_of_edges:
+                break
+    edges_below = {}
+    edge = 0
+    while edge < number_of_edges:
+        parent = tables.edges[edge].parent
+        edges_below[parent] = [edge]
+        edge += 1
+        if edge == number_of_edges:
+            break
+        while tables.edges[edge].parent == parent:
+            edges_below[parent].append(edge)
+            edge += 1
+            if edge == number_of_edges:
+                break
+    number_of_mutations = len(tables.mutations)
+    if theta == 0:
+        if number_of_mutations == 0:
+            ret = 0
+        else:
+            ret = -float("inf")
+    else:
+        ret = (number_of_mutations * math.log(total_material * theta) -
+               total_material * theta)
+    for mut in arg.mutations():
+        mutant_location = tables.sites[mut.site].position
+        ind = 0
+        edge = edges_above[mut.node][0]
+        while tables.edges[edge].right < mutant_location:
+            ind += 1
+            edge = edges_above[mut.node][ind]
+        parent = tables.edges[edge].parent
+        child = tables.edges[edge].child
+        potential_branch_length = tables.nodes[parent].time - tables.nodes[child].time
+        continue_downwards = True
+        if tables.nodes[child].time == 0:
+            continue_downwards = False
+        tmp_edge = edge
+        while continue_downwards:
```

Comment:
Can you explain in words what this is doing? I don't quite get it, and I think it can be made a bit more concise.

Reply:
Each infinite sites mutation can appear on one of a set of contiguous edges, and I'm assuming it has been recorded on an arbitrary valid edge. The job of the `while continue_downwards` and `while continue_upwards` loops starting on respective lines 93 and 116 is to find the extremal valid edges; more specifically, the total amount of branch length in between. Continue_downwards goes towards the leaves, continue_upwards towards the root. I suppose that already suggests better variable names for the next update.

Going leafwards, the valid window of edges ends is if the current child node is the parent of a coalescence event, and both of its children carry ancestral material at the mutant site. That means the mutation couldn't have happened any further leafward, while still appearing on all necessary leaves. The variable `count` counts how many leafward edges carry ancestral material, and we stop looking when it fails to equal one. The amount of available branch length for the mutation gets stored in `potential_branch_length`.

Going rootwards, the window of valid edges can end in two ways. The current parent node could have two offspring (one of which is the current child node), both of which carry ancestral material at the mutant site (analogously to how the leafward case terminated). Alternatively, no edge above the current parent node carries ancestral material at the mutant site, in which case it has reached its MRCA. Now that I write that out, I realise that the former condition is implied by the latter, so this can indeed be simplified. As it is, the variable `count_down` checks the former condition, and `count_up` the latter.

Reply:
OK, that's helpful, thanks. I think another iteration where we store the edge objects themselves in the lookup dicts will help to clear things up a little. Would you mind making these changes please?

Reply:
Not at all. In fact, many of them are already made. I'll read up on defaultdict and see if it's cleverer than I'm giving it credit for, and then upload a new version.